frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -208,13 +209,74 @@ private void refreshModel(GitRepository gitRepository, List<String> skippedBranc
   }
 
   @UIEffect
-  private void openDiscoverDialog() {
-    var action = ActionManager.getInstance().getAction(ACTION_DISCOVER);
-    var dataContext = DataManager.getInstance().getDataContext(EnhancedGraphTable.this);
-    var anActionEvent = AnActionEvent.createFromDataContext(ACTION_PLACE_EMPTY_TABLE, new Presentation(), dataContext);
-    action.actionPerformed(anActionEvent);
+  private void doAutomaticDiscover(Path macheteFilePath) {
+    var selectedRepository = getGitRepositorySelectionProvider().getSelectedGitRepository().getOrNull();
+    if (selectedRepository == null) {
+      LOG.error("Can't do automatic discover because of undefined selected repository");
+      return;
+    }
+    var mainDirPath = GitVfsUtils.getMainDirectoryPath(selectedRepository).toAbsolutePath();
+    var gitDirPath = GitVfsUtils.getGitDirectoryPath(selectedRepository).toAbsolutePath();
+
+    @SuppressWarnings({"nullness", "guieffect"}) // To allow runResult be uninitialized and Tuple accept @Nullable IBranchLayout
+    var task = new Task.Backgroundable(project,
+        getString("string.GitMachete.EnhancedGraphTable.automatic-discover.task-title")) {
+      private Try<Tuple3<IBranchLayoutWriter, @Nullable IBranchLayout, IGitMacheteRepositorySnapshot>> runResult;
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        runResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot())
+            .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
+                __ -> new EmptyRepositoryException())
+            .map(repositorySnapshot -> {
+              var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+              var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+              return Tuple(branchLayoutWriter, branchLayout, repositorySnapshot);
+            })
+            .filterTry(tuple -> tuple._2 != null, __ -> new EmptyBranchLayoutException())
+            .andThenTry(tuple -> tuple._1.write(macheteFilePath, tuple._2, /* backupBranchLayout */ true));
+      }
+      // This method is invoked on UI dispatch thread
+      @UIEffect
+      @Override
+      public void onSuccess() {
+        // Match below returns empty string just because must return something
```

Comment:
Hmm TBH `Match` feels a bit artificial here (also due to the reason stated in this comment)... anyway, maybe it's not a good idea to handle the failures in `onSuccess` callback? Can we use `onFailure` or a similar callback for that purpose?

Reply:
Unfortunately with `Task.Backgroundable` not. I try use a `Task.WithResult` but handling exception when receive the result is awkward (this is how it works in that case).

Reply:
Out of date. Whole job is done in `run` method.